{{cookiecutter.package_name}}/src/{{cookiecutter.package_name}}/data/builder.py

```diff
@@ -0,0 +1,129 @@
+
+   Logging in this module should be done at the ``debug`` level.
+
+"""
+from pathlib import Path
+
+from loguru import logger
+import pandas as pd
+from vivarium.framework.artifact import Artifact, get_location_term, EntityKey
+
+from {{cookiecutter.package_name}} import globals as project_globals
+from {{cookiecutter.package_name}}.data import loader
+
+
+def open_artifact(output_path: Path, location: str) -> Artifact:
+    """Creates or opens an artifact at the output path.
+
+    Parameters
+    ----------
+    output_path
+        Fully resolved path to the artifact file.
+    location
+        Proper GBD location name represented by the artifact.
+
+    Returns
+    -------
+        A new artifact.
+
+    """
+    if not output_path.exists():
+        logger.debug(f"Creating artifact at {str(output_path)}.")
+    else:
+        logger.debug(f"Opening artifact at {str(output_path)} for appending.")
+
+    artifact = Artifact(output_path, filter_terms=[get_location_term(location)])
+
+    key = project_globals.METADATA_LOCATIONS
+    if key not in artifact:
+        artifact.write(key, [location])
+
+    return artifact
+
+
+def load_and_write_data(artifact: Artifact, key: str, location: str):
+    """Loads data and writes it to the artifact if not already present.
+
+    Parameters
+    ----------
+    artifact
+        The artifact to write to.
+    key
+        The entity key associated with the data to write.
+    location
+        The location associated with the data to load and the artifact to
+        write to.
+
+    """
+    if key in artifact:
+        logger.debug(f'Data for {key} already in artifact.  Skipping...')
+    else:
+        logger.debug(f'Loading data for {key} for location {location}.')
+        data = loader.get_data(key, location)
+        logger.debug(f'Writing data for {key} to artifact.')
+        artifact.write(key, data)
+    return artifact.load(key)
+
+
+def write_data(artifact: Artifact, key: str, data: pd.DataFrame):
+    """Writes data to the artifact if not already present.
+
+    Parameters
+    ----------
+    artifact
+        The artifact to write to.
+    key
+        The entity key associated with the data to write.
+    data
+        The data to write.
+
+    """
+    if key in artifact:
+        logger.debug(f'Data for {key} already in artifact.  Skipping...')
+    else:
+        logger.debug(f'Writing data for {key} to artifact.')
+        artifact.write(key, data)
+    return artifact.load(key)
+
+
+def write_data_by_draw(artifact: Artifact, key: str, data: pd.DataFrame):
+    with pd.HDFStore(artifact.path, complevel=9, mode='a') as store:
+        key = EntityKey(key)
+        artifact._keys.append(key)
+        store.put(f'{key.path}/index', data.index.to_frame(index=False))
+        data = data.reset_index(drop=True)
+        for c in data.columns:
+            store.put(f'{key.path}/{c}', data[c])
+
+
+def load_and_write_demographic_data(artifact: Artifact, location: str):
```

Comment:
I think it's okay for these functions which just collect and write keys to live without docstrings